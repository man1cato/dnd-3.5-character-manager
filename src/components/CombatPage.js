import React from 'react'
import { connect } from 'react-redux'
import update from 'immutability-helper'

import Header from './Header'
import Footer from './Footer'
import PhysicalStats from './PhysicalStats'
import Saves from './Saves'
import Attacks from './Attacks'
import Weapons from './Weapons'
import PreparedSpells from './PreparedSpells'
import { startEditProfile } from '../actions/profile'


const setNestedProps = (prop) => {
	Object.keys(prop).forEach((key) => {
		prop[key].mod = prop[key].mod || ""
		prop[key].total = prop[key].total || prop[key].base
	})
	return prop
}

export class CombatPage extends React.Component {
	state = {
		hp: {
			base: this.props.hp.base,
			mod: this.props.hp.mod || "",
			damage: this.props.hp.damage || "",
			total: this.props.hp.total || this.props.hp.base,
		},
		initiative: {
			base: this.props.initiative.base, 
			mod: this.props.initiative.mod || "",
			total: this.props.initiative.total || this.props.initiative.base
		},
		saves: setNestedProps(this.props.saves),      
		attacks: setNestedProps(this.props.attacks)
	}  	

	handleChange = (e) => {
		const name = e.target.name
		const id = e.target.id
		let value = Number(e.target.value)
		value = (value === 0 || isNaN(value)) ? "" : value
		let total
		this.setState((prevState) => {
			if (name === "hp") {
				if (id === "damage") {
					total = this.props.hp.base + prevState.hp.mod - value
					return {
						hp: update(prevState.hp, {
						damage: { $set: value },
						total: { $set: total }
						})
					}
				} 
				total = this.props.hp.base + value - prevState.hp.damage
			} else if (name === "saves" || name === "attacks") {
				total = this.props[name][id].base + value;
				return {
					[name]: update(prevState[name], {
						[id]: {
						mod: { $set: value },
						total: { $set: total }
						}
					})
				}
			} else {
				total = this.props[name].base + value
			}
			return {
				[name]: update(prevState[name], {
					mod: { $set: value },
					total: { $set: total }          
				})
			}
		}, () => {
			this.props.startEditProfile(this.props.id, { [name]: this.state[name] })
		});
	}

	render () {
		return (
			<div className="layout">
				<Header pageTitle="Combat" />
				<div className="container container--body">

					<PhysicalStats
						hp={this.state.hp}
						ac={this.props.ac}
						initiative={this.state.initiative}
						speed={this.props.speed}
						handleChange={this.handleChange}
					/>                              

					<Saves
						saves={this.state.saves}            
						handleChange={this.handleChange}
					/>            

					<Attacks 
						attacks={this.state.attacks}
						handleChange={this.handleChange}
					/>

					<Weapons 
						weaponSet={this.props.weaponSet}
						meleeBonus={this.state.attacks.melee.total}
						rangedBonus={this.state.attacks.ranged.total}
						grappleBonus={this.state.attacks.grapple.total}
					/>

					{this.props.spellbook && 
						<PreparedSpells 
							id={this.props.id}
							spellbook={this.props.spellbook}
							spells={this.props.spells}
							startEditProfile={this.props.startEditProfile}
						/>
					}

				</div>

				<Footer />
			</div>  
		)
	}
}

const mapStateToProps = (state) => ({
  id: state.profile.id,
  hp: state.profile.hp,
  speed: state.profile.speed,
  ac: state.profile.ac,
  initiative: state.profile.initiative,
  saves: state.profile.saves,
  bab: state.profile.bab,
  attacks: state.profile.attacks,
  weaponSet: state.profile.weaponSet,
  spellbook: state.profile.spellbook,
  spells: state.spells
})

const mapDispatchToProps = (dispatch, props) => ({
  startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
});


export default connect(mapStateToProps, mapDispatchToProps)(CombatPage)