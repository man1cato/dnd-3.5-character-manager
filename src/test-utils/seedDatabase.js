import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}