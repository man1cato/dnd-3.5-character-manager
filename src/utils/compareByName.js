export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}