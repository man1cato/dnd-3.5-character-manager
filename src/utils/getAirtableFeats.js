import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}