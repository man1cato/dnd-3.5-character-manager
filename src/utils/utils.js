import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])