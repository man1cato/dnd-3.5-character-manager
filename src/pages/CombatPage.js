import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import PhysicalStats from '../components/PhysicalStats'
import AttackBonuses from '../components/AttackBonuses/AttackBonuses'
import Saves from '../components/Saves/Saves'
import Weapons from '../components/Weapons'
import PreparedSpells from '../components/PreparedSpells'
import { startEditProfile } from '../store/actions/profile'
import { calcAbilityMod } from '../utils/utils'


export const CombatPage = ({ profile, races, jobClasses, spells, startEditProfile }) => {
	const jobClassLevel = jobClasses[profile.jobClass].levels[profile.level]
	const race = races[profile.race]
	const handleUpdate = (updates) => {	startEditProfile(profile.id, updates) }

	return (
		<div className="container container--body">
			<div className="section">
				<h3 className="section__title">Physical Stats</h3>
				<PhysicalStats
					hp={profile.hp}
					ac={profile.ac}
					initBase={
						!!profile.abilities.dex.tempScore ? 
						calcAbilityMod(profile.abilities.dex.tempScore) : 
						calcAbilityMod(profile.abilities.dex.score)
					}
					initMod={profile.initMod}
					speed={race.speed}
					handleUpdate={handleUpdate}							
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Saving Throws</h3>
				<Saves
					saveMods={profile.saveMods}
					saveBases={jobClassLevel.saves}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">       
				<h3 className="section__title">Attack Bonuses</h3>
				<AttackBonuses 
					attackBonusMods={profile.attackBonusMods}
					abilities={profile.abilities}
					baseAttackBonuses={jobClassLevel.baseAttackBonuses}
					size={race.size}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Equipped Weapons</h3>
				<Weapons 
					weaponSet={profile.equipped.weapons}
					// meleeBonus={state.attacks.melee.total}
					// rangedBonus={state.attacks.ranged.total}
					// grappleBonus={state.attacks.grapple.total}
				/>
			</div>

			{profile.spellbook && (
				<div className="section">
					<h3 className="section__title">Prepared Spells</h3>
					<PreparedSpells 
						spellbook={profile.spellbook}
						spells={spells}
						handleUpdate={handleUpdate}
					/>
				</div>
			)}

		</div>
	)
}

const mapStateToProps = ({ profile, races, jobClasses, spells }) => ({
	profile,
	races,
	jobClasses,
	spells
})

const mapDispatchToProps = (dispatch) => ({
  	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})


export default connect(mapStateToProps, mapDispatchToProps)(CombatPage)