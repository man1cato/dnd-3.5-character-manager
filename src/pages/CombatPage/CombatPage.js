import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import PhysicalStats from '../../components/PhysicalStats/PhysicalStats'
import AttackBonuses from '../../components/AttackBonuses/AttackBonuses'
import Saves from '../../components/Saves/Saves'
import Weapons from '../../components/Weapons/Weapons'
import PreparedSpells from '../../components/PreparedSpells/PreparedSpells'
import { startEditProfile } from '../../store/actions/profile'
import { calcAbilityMod } from '../../utils/utils'


export const CombatPage = ({ profile, races, jobClasses, spells, startEditProfile }) => {
	const jobClassLevel = jobClasses[profile.jobClass].levels[profile.level]
	const race = races[profile.race]
	const handleUpdate = updates => { 
		startEditProfile(profile.id, updates) 
	}

	return (
		<div className="container--body">
			<div className="section">
				<h3 className="section__title">Physical Stats</h3>
				<PhysicalStats
					hp={profile.hp}
					ac={profile.ac}
					initBase={
						!!profile.abilities.dex.tempScore ? 
						calcAbilityMod(profile.abilities.dex.tempScore) : 
						calcAbilityMod(profile.abilities.dex.score)
					}
					initMod={profile.initiativeMod}
					speed={race.speed}
					handleUpdate={handleUpdate}							
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Saving Throws</h3>
				<Saves
					saveMods={profile.saveMods}
					saveBases={jobClassLevel.saves}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">       
				<h3 className="section__title">Attack Bonuses</h3>
				<AttackBonuses 
					attackBonuses={profile.attackBonuses}
					handleUpdate={handleUpdate}
				/>
			</div>

			{profile.equipped && ( 
				<div className="section">
					<h3 className="section__title">Equipped Weapons</h3>
					<Weapons 
						equippedWeapons={profile.equipped.weapons}
						attackBonuses={profile.attackBonuses}
					/>
				</div>
			)}

			{profile.spellbook && (
				<div className="section">
					<h3 className="section__title">Prepared Spells</h3>
					<PreparedSpells 
						spellbook={profile.spellbook}
						spells={spells}
						handleUpdate={handleUpdate}
					/>
				</div>
			)}

		</div>
	)
}

const mapStateToProps = state => ({
	profile: state.profile,
	races: state.races,
	jobClasses: state.jobClasses,
	spells: state.spells
})

const mapDispatchToProps = dispatch => ({
  	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})


export default connect(mapStateToProps, mapDispatchToProps)(CombatPage)