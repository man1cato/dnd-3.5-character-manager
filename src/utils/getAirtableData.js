import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}