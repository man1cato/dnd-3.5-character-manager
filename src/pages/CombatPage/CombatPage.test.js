import React from 'react'
import { shallow } from 'enzyme'
import { render, fireEvent } from '@testing-library/react'

import { CombatPage } from './CombatPage'
import { characterOne } from '../../tests/utils/seedDatabase'
import { apiData } from '../../tests/utils/utils'
 

const startEditProfile = jest.fn()
let props

beforeAll(async () => {
    const api = await apiData()
    const jobClasses = api.jobClasses
    props = {
        profile: characterOne,
        races: api.races,
        jobClasses,
        spells: api.spells,
        startEditProfile
    }
})


test('should render combat page with profile data', () => {
    const wrapper = shallow(<CombatPage {...props} />)
    expect(wrapper).toMatchSnapshot()
})