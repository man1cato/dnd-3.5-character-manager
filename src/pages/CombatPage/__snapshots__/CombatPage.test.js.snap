// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render combat page with profile data 1`] = `
<div>
  <div
    class="container--body"
  >
    <div
      class="section"
    >
      <h3
        class="section__title"
      >
        Physical Stats
      </h3>
      <div
        class="PhysicalStats__row"
      >
        <div>
          Hit Points
        </div>
        <div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Base
            </h5>
            <div>
              32
            </div>
          </div>
          <div
            class="PhysicalStats__row__item--bottom"
          >
            +
          </div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Mod
            </h5>
            <input
              data-testid="hpMod"
              type="number"
              value="6"
            />
          </div>
          <div
            class="PhysicalStats__row__item--bottom"
          >
            -
          </div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Dmg
            </h5>
            <input
              data-testid="hpDamage"
              type="number"
              value="10"
            />
          </div>
          <div
            class="PhysicalStats__row__item--bottom"
          >
            =
          </div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Curr
            </h5>
            <div
              data-testid="hpTotal"
            >
              28
            </div>
          </div>
        </div>
      </div>
      <div
        class="PhysicalStats__row"
      >
        <div>
          Armor Class
        </div>
        <div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Base
            </h5>
            <div>
              14
            </div>
          </div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Flat
            </h5>
            <div>
              10
            </div>
          </div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Touch
            </h5>
            <div>
              14
            </div>
          </div>
        </div>
      </div>
      <div
        class="PhysicalStats__row"
      >
        <div>
          Initiative
        </div>
        <div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Base
            </h5>
            <div>
              4
            </div>
          </div>
          <div
            class="PhysicalStats__row__item--bottom"
          >
            +
          </div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Mod
            </h5>
            <input
              data-testid="initMod"
              type="number"
              value="2"
            />
          </div>
          <div
            class="PhysicalStats__row__item--bottom"
          >
            =
          </div>
          <div
            class="PhysicalStats__row__item"
          >
            <h5>
              Total
            </h5>
            <div>
              6
            </div>
          </div>
        </div>
      </div>
      <div
        class="PhysicalStats__row"
      >
        <div>
          Speed
        </div>
        <div>
          30
           ft/round
        </div>
      </div>
    </div>
    <div
      class="section"
    >
      <h3
        class="section__title"
      >
        Saving Throws
      </h3>
      <div
        class="Saves"
      >
        <h5
          class="grid__col1"
        >
          Type
        </h5>
        <h5>
          Base
        </h5>
        <div />
        <h5>
          Mod
        </h5>
        <div />
        <h5>
          Total
        </h5>
        <div
          class="grid__col1"
        >
          Fortitude
        </div>
        <div>
          1
        </div>
        <div>
          +
        </div>
        <input
          data-testid="fortitudeMod"
          type="number"
          value="1"
        />
        <div>
          =
        </div>
        <div
          data-testid="fortitudeTotal"
        >
          2
        </div>
        <div
          class="grid__col1"
        >
          Reflex
        </div>
        <div>
          1
        </div>
        <div>
          +
        </div>
        <input
          data-testid="reflexMod"
          type="number"
          value="-1"
        />
        <div>
          =
        </div>
        <div
          data-testid="reflexTotal"
        >
          0
        </div>
        <div
          class="grid__col1"
        >
          Will
        </div>
        <div>
          4
        </div>
        <div>
          +
        </div>
        <input
          data-testid="willMod"
          type="number"
          value=""
        />
        <div>
          =
        </div>
        <div
          data-testid="willTotal"
        >
          4
        </div>
      </div>
    </div>
    <div
      class="section"
    >
      <h3
        class="section__title"
      >
        Attack Bonuses
      </h3>
      <div
        class="AttackBonuses"
      >
        <h5
          class="grid__col1"
        >
          Type
        </h5>
        <h5>
          Base
        </h5>
        <div />
        <h5>
          Mod
        </h5>
        <div />
        <h5>
          Total
        </h5>
        <div
          class="grid__col1"
        >
          Melee
        </div>
        <div>
          4
        </div>
        <div>
          +
        </div>
        <input
          data-testid="meleeMod"
          type="number"
          value="2"
        />
        <div>
          =
        </div>
        <div>
          6
        </div>
        <div
          class="grid__col1"
        >
          Ranged
        </div>
        <div>
          6
        </div>
        <div>
          +
        </div>
        <input
          data-testid="rangedMod"
          type="number"
          value=""
        />
        <div>
          =
        </div>
        <div>
          6
        </div>
        <div
          class="grid__col1"
        >
          Grapple
        </div>
        <div>
          4
        </div>
        <div>
          +
        </div>
        <input
          data-testid="grappleMod"
          type="number"
          value="1"
        />
        <div>
          =
        </div>
        <div>
          5
        </div>
      </div>
    </div>
    <div
      class="section"
    >
      <h3
        class="section__title"
      >
        Equipped Weapons
      </h3>
      <div
        class="weapon"
      >
        <h4>
          Repeating Light Crossbow
        </h4>
        <div
          class="weapon__features"
        >
          Exotic
          , 
          Ranged
          , 
          Piercing
        </div>
        <div
          class="weapon__stats"
        >
          <div>
            <h5>
              Rng
            </h5>
            <div>
              80 ft.
            </div>
          </div>
          <div>
            <h5>
              Atk
            </h5>
            <div>
              +
              6
            </div>
          </div>
          <div>
            <h5>
              Dmg
            </h5>
            <div>
              1d8
            </div>
          </div>
          <div>
            <h5>
              Crit
            </h5>
            <div>
              19–20/×2
            </div>
          </div>
        </div>
      </div>
      <div
        class="weapon"
      >
        <h4>
          Quarterstaff
        </h4>
        <div
          class="weapon__features"
        >
          Simple
          , 
          Two-Handed
          , 
          Bludgeoning
        </div>
        <div
          class="weapon__stats"
        >
          <div>
            <h5>
              Rng
            </h5>
            <div>
              —
            </div>
          </div>
          <div>
            <h5>
              Atk
            </h5>
            <div>
              +
              6
            </div>
          </div>
          <div>
            <h5>
              Dmg
            </h5>
            <div>
              1d6/1d6
            </div>
          </div>
          <div>
            <h5>
              Crit
            </h5>
            <div>
              ×2
            </div>
          </div>
        </div>
      </div>
      <div
        class="weapon"
      >
        <h4>
          Light Crossbow
        </h4>
        <div
          class="weapon__features"
        >
          Simple
          , 
          Ranged
          , 
          Piercing
        </div>
        <div
          class="weapon__stats"
        >
          <div>
            <h5>
              Rng
            </h5>
            <div>
              80 ft.
            </div>
          </div>
          <div>
            <h5>
              Atk
            </h5>
            <div>
              +
              6
            </div>
          </div>
          <div>
            <h5>
              Dmg
            </h5>
            <div>
              1d8
            </div>
          </div>
          <div>
            <h5>
              Crit
            </h5>
            <div>
              19–20/×2
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="section"
    >
      <h3
        class="section__title"
      >
        Prepared Spells
      </h3>
      <div
        class="PreparedSpells__grid"
      >
        <h5
          class="grid__col1"
        >
          Spell
        </h5>
        <h5
          class="grid__col2"
        >
          Rmng
        </h5>
        <button
          class="grid__col1 button--link"
          data-testid="rec8IQeYmBr3f1BMC"
          type="button"
        >
          Acid Splash
        </button>
        <div
          class="grid__col2"
          data-testid="rec8IQeYmBr3f1BMCRemaining"
        >
          1
        </div>
        <button
          change="1"
          class="grid__col4"
          data-testid="rec8IQeYmBr3f1BMCCastButton"
          level="0"
          spellid="rec8IQeYmBr3f1BMC"
        >
          Cast
        </button>
        <button
          class="grid__col1 button--link"
          data-testid="recwpeRoCsea4vidW"
          type="button"
        >
          Arcane Mark
        </button>
        <div
          class="grid__col2"
          data-testid="recwpeRoCsea4vidWRemaining"
        >
          1
        </div>
        <button
          change="1"
          class="grid__col4"
          data-testid="recwpeRoCsea4vidWCastButton"
          level="0"
          spellid="recwpeRoCsea4vidW"
        >
          Cast
        </button>
        <button
          class="grid__col1 button--link"
          data-testid="recRhVs0ahtQZpbXx"
          type="button"
        >
          Charm Person
        </button>
        <div
          class="grid__col2"
          data-testid="recRhVs0ahtQZpbXxRemaining"
        >
          0
        </div>
      </div>
    </div>
  </div>
</div>
`;
